task: introduce sftp

benjaminkott opened the pull request:
This patch introduces Flysystem (https://flysystem.thephpleague.com/) as
the underlying package to handle remote storage. For now, we are only
integrating SFTP in addition to FTP to allow more connection options for
FTP uploads. There will now be an option to select FTP or SFTP, and the
correct adapter will be used to create the connection.

The code has been restructured to move logic away from the endpoints
towards a central service. This will make it easier to implement and
maintain in the future.

Some options have been removed as they do not seem to work reliably:

- `ftp:appendDate` is a nice idea, but it has a flawed implementation
  as the date is evaluated on each request. This results in a mismatch
  between requested and actual URLs to files.

- `ftp:upload_thumb`: Thumbnails are now always uploaded, ensuring all
  files are available on the server for the generated webpage.
  Defaults were introduced for several config settings. This way, we can
  rely on the settings being set and do not need to guess.

The template for the server has been restructured. Instead of parsing
the template and potentially risking breaking it, we are now providing
a dedicated config file that can be loaded without problems.
Additionally, a test template that uses local data is now available.

andi34 commented on src/Utility/PathUtility.php:
```diff
@@ -23,10 +23,6 @@ public static function getAbsolutePath(string $path = ''): string
             return '';
         }
 
-        if (self::isUrl($path)) {
```
We need this to allow fonts and backgrounds being used via URL

benjaminkott replied:
@andi34 this destroys the purpose of the function, i need to be able to rely on getting an absolute path. not getting an absolute path here makes it useless.

andi34 replied:
Well, then we need a different function for config defined backgrounds, placeholder and fonts to check for the url first

benjaminkott replied:
will check the usages, but i don't think that's useful to have fonts, and images stored remotely that need to be processed locally - all other places that are only displaying things in the frontend do not need to have these checks. we offer to upload fonts, and images to the system through the interface, i don´t see a use case of having them stored remotely as it will slow down the system.

benjaminkott replied:
even the current placement of the ftp upload drives me crazy, because the box stops working when the file can not be uploaded. normally you store the upload requests in a dedicated queue - that will be processed through a dedicated background process to avoid slowing down the system. not sure it i can manage to bring this to the first draft here to extend remote storage capabilities as i need to have this working in about 2 weeks.

benjaminkott replied:
i could not spot a problem here, so i am not sure what problem is fixed with the url hack.

benjaminkott commented:
@fmiccolis as you did the original implementation, please check the updated code - I've made some adjustments for the better i guess, as it´s a bit more reliable and adjustable.

fmiccolis commented:
@benjaminkott ok I'll test the new code!

fmiccolis commented:
Hi @benjaminkott ,
the "old" configuration works fine. The files are uploaded and deleted as before.
I don't have a SFTP server to use to test the new part so I cannot help with it.

I was thinking about something: in the file uploader we can let the user upload it's own template for the webpage by adding the folder /template, what do you think?

benjaminkott commented:
@fmiccolis 

> I was thinking about something: in the file uploader we can let the user upload its own template for the webpage by adding the folder /template, what do you think?

I am still unsure about this one, as it requires some kind of expert knowledge. 

If you know what you are doing - it should be more or less easy to adjust. Everyone else, should probably only use the provided option.

Did you remove the initial created `index.php` file and check for the new generated template? I checked yesterday with some collegues and we needed to adjust some stuff to get it working.

fmiccolis commented:
> If you know what you are doing - it should be more or less easy to adjust. Everyone else, should probably only use the provided option.

Yes, maybe You're right..


> Did you remove the initial created index.php file and check for the new generated template? I checked yesterday with some collegues and we needed to adjust some stuff to get it working.

I have changed the title so in my ftp server a new folder has been created and the template loaded on.
The template is different from the old one and uses the configuration file uploaded with the index.php template file.
I don't know if I responded to your question.

benjaminkott committed:
bugfix: use HTTP_HOST instead of SERVER_NAME

benjaminkott committed:
task: introduce sftp

This patch introduces Flysystem (https://flysystem.thephpleague.com/) as
the underlying package to handle remote storage. For now, we are only
integrating SFTP in addition to FTP to allow more connection options for
FTP uploads. There will now be an option to select FTP or SFTP, and the
correct adapter will be used to create the connection.

The code has been restructured to move logic away from the endpoints
towards a central service. This will make it easier to implement and
maintain in the future.

Some options have been removed as they do not seem to work reliably:

- `ftp:appendDate` is a nice idea, but it has a flawed implementation
  as the date is evaluated on each request. This results in a mismatch
  between requested and actual URLs to files.

- `ftp:upload_thumb`: Thumbnails are now always uploaded, ensuring all
  files are available on the server for the generated webpage.
  Defaults were introduced for several config settings. This way, we can
  rely on the settings being set and do not need to guess.

The template for the server has been restructured. Instead of parsing
the template and potentially risking breaking it, we are now providing
a dedicated config file that can be loaded without problems.
Additionally, a test template that uses local data is now available.

Change-Id: Iac24b838617c6d1a4147a385236cd430cad3eb09

benjaminkott committed:
bugfix: sort images and check for existing thumbnails

Change-Id: Ie21fde75c51c605600a407e6c08de33c8834576c
